File: built-in plugins dsp/mb-auto-filter.cpp
/**
 * MB Auto Filter
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Resonant filter with envelope follower and LFO modulation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_AUTO_FILTER_H
#define MB_AUTO_FILTER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbAutoFilter : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-auto-filter";
    static constexpr const char* PLUGIN_NAME    = "MB Auto Filter";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float cutoff = 1000f;  // range [20, 20000]
    float resonance = 0.5f;  // range [0, 1]
    float env_amount = 0f;  // range [-100, 100]
    float env_attack = 10f;  // range [0.1, 500]
    float env_release = 100f;  // range [1, 2000]
    float lfo_rate = 1f;  // range [0.01, 20]
    float lfo_amount = 0f;  // range [0, 100]
    float drive = 0f;  // range [0, 1]
    };

    MbAutoFilter() = default;
    ~MbAutoFilter() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.cutoff = std::clamp(params.cutoff, 20f, 20000f);
        params.resonance = std::clamp(params.resonance, 0f, 1f);
        params.env_amount = std::clamp(params.env_amount, -100f, 100f);
        params.env_attack = std::clamp(params.env_attack, 0.1f, 500f);
        params.env_release = std::clamp(params.env_release, 1f, 2000f);
        params.lfo_rate = std::clamp(params.lfo_rate, 0.01f, 20f);
        params.lfo_amount = std::clamp(params.lfo_amount, 0f, 100f);
        params.drive = std::clamp(params.drive, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Auto Filter
        return input;
    }
};

#endif // MB_AUTO_FILTER_H

File: built-in plugins dsp/mb-auto-filter.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbAutoFilterPlugin: PluginDefinition = {
    id: 'mb-auto-filter',
    slug: 'mb-auto-filter',
    name: 'MB Auto Filter',
    category: 'effect',
    type: 'eq',
    version: '1.0.0',
    description: 'Resonant filter with envelope follower and LFO modulation',
    author: 'Max Booster',
    parameters: [
      { id: 'filter_type', name: 'Filter Type', type: 'choice', defaultValue: 'lowpass', choices: ['lowpass', 'highpass', 'bandpass', 'notch'], automatable: false },
      { id: 'cutoff', name: 'Cutoff', type: 'float', defaultValue: 1000, minValue: 20, maxValue: 20000, unit: 'Hz', automatable: true },
      { id: 'resonance', name: 'Resonance', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true },
      { id: 'env_amount', name: 'Envelope Amount', type: 'float', defaultValue: 0, minValue: -100, maxValue: 100, unit: '%', automatable: true },
      { id: 'env_attack', name: 'Env Attack', type: 'float', defaultValue: 10, minValue: 0.1, maxValue: 500, unit: 'ms', automatable: true },
      { id: 'env_release', name: 'Env Release', type: 'float', defaultValue: 100, minValue: 1, maxValue: 2000, unit: 'ms', automatable: true },
      { id: 'lfo_rate', name: 'LFO Rate', type: 'float', defaultValue: 1, minValue: 0.01, maxValue: 20, unit: 'Hz', automatable: true },
      { id: 'lfo_amount', name: 'LFO Amount', type: 'float', defaultValue: 0, minValue: 0, maxValue: 100, unit: '%', automatable: true },
      { id: 'lfo_shape', name: 'LFO Shape', type: 'choice', defaultValue: 'sine', choices: ['sine', 'triangle', 'square', 'sawtooth', 'random'], automatable: false },
      { id: 'drive', name: 'Drive', type: 'float', defaultValue: 0, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { filter_type: 'lowpass', cutoff: 1000, resonance: 0.5, env_amount: 0, env_attack: 10, env_release: 100, lfo_rate: 1, lfo_amount: 0, lfo_shape: 'sine', drive: 0 },
  };

export default MbAutoFilterPlugin;
